Remove molecule from the package building process

legoktm opened the pull request:
## Status

Ready for review

## Description of Changes

Split into 5 commits:
* ~~Pin dh-virtualenv from jammy, not impish~~ Cherry-picked into https://github.com/freedomofpress/securedrop/pull/6755.
* Remove not-useful builder tests (see commit message for detailed rationale)
* Split and slim down test_securedrop_deb_package.py (again, see commit message)
* Re-organize ossec packaging to use debhelper and standard layout (basically the equivalent of https://github.com/freedomofpress/securedrop/pull/6544 but for OSSEC packages)
* Use plain container for package building, not molecule
* Verify all packages except app-code are reproducible in CI (see commit message for detailed rationale)

Fixes #6706.

## Testing

* [x] `make build-debs` ~without this PR~ from `0fb3f24^`, save the debs somewhere ("OG packages")
* [x] `make build-debs && make build-debs-ossec` with this PR ("Round 1 packages")
* [x] diffoscope the old and new packages. Expectations:
   * [x] securedrop-app-code: no non-trivial diffs (compiled .so files will be different, maybe some timestamp variation that is being fixed AFAICT)
   * [x] securedrop-config: should have identical sha256 checksum
   * [x] securedrop-keyring: should have identical sha256 checksum
   * [x] securedrop-ossec-{agent,server}: should have identical sha256 checksum
   * [x] ossec-{agent,server}: honestly it's going to be a mess because all the compiled files are different and the timestamps were previously not reproducible and now they are.
      * It's probably worth just installing these on a prod setup and verifying the notifications get sent properly as far as testing goes.
* [x] Save the ossec-{agent,server} debs from "Round 1" somewhere, then run `make build-debs-ossec` again ("Round 2"). Verify that the sha256 checksums are identical from "Round 1" and "Round 2". Aka the new ossec packages are fully bit-for-bit reproducible.
* [x] CI is passing

## Deployment
My expectation is that we're not actually going to deploy the new ossec-{agent,server} debs because it's roughly the same as the existing 3.6.0 debs. So we can use diffoscope to verify confidence in the main securedrop packages that nothing meaningful is changing.

## Checklist

- [x] [Configuration tests](https://docs.securedrop.org/en/latest/development/testing_configuration_tests.html) pass
- [x] I have written a test plan and validated it for this PR
- [x] I have proposed documentation updates: https://github.com/freedomofpress/securedrop-dev-docs/pull/40

legoktm committed:
Remove not longer useful builder tests

test_build_dependencies.py:
  The two non-disabled tests just assert the Python 3 / dh_virtualenv
version, which can be done statically by examining the build container
  rather than needing tests for it.

test_legacy_paths.py:
  These should be gone, everyone has surely reset their build
  environment since those changes were made.

legoktm committed:
Split and slim down test_securedrop_deb_package.py

In preparation for splitting the OSSEC builds out of the main SecureDrop
package builds, split the tests. Since we want to get away from molecule
eventually, remove the use of testinfra and yaml variables.

I also removed tests that I didn't feel were useful or were already
covered at a higher level. In general for non-critical things (e.g. the
maintainer address) we trust that debhelper/dpkg did the right thing and
catch regressions via debdiff/diffoscope of previous versions.

* test_deb_package_control_fields: Can be statically verified by reading
  d/control.
* test_deb_package_control_fields_homepage: Same.
* test_securedrop_app_code_contains_css: CSS is no longer generated on
  the fly, so we can assume that it will be copied properly, like every
  other file.
* test_securedrop_app_code_contains_static_fonts and co.: Same.
* test_control_helper_files_are_present: Trust that debhelper installs
  the maintainer scripts properly.

The various tests that check for existence or absence of specifc files
have been consolidated into two parameterized tests.

legoktm committed:
Re-organize ossec packaging to use debhelper and standard layout

Same as the reorganization of the securedrop packages, use a normal
debian/ folder and have debhelper install files into the package.

Hardcode the remaining jinja2 variables and drop the obsolete init.d
scripts.

legoktm committed:
Use plain container for package building, not molecule

Simplify the package building process by using plain containers that
directly execute `dpkg-buildpackage` with minimal layers of indirection.
Some of the $OCI_BIN, etc. logic is copied or inspired from the
container scripts in securedrop-updater.

The container is explicitly not pinned, instead it's built locally and
if any packages are out of date, rebuilt, before any package building
begins. It's mostly the same, the main difference is that we warm the
cargo cache so it's not fetched cold on each build.

From a developer perspective, it's now trivial to see the build output,
which incidentally reveals some warnings we've been ignoring. OSSEC packages
are also no longer built by default and have separate make targets for
when we want to actually update those.

CI will now build the debs on each commit and PR since it's fast enough
now.

Fixes #6706.

legoktm commented:
Discussed with Kev today, one thing to add either here or in documentation is what should be done with buildinfo files (see #6356) and investigate setting up a rebuilder to independently verify reproducibility.

legoktm commented:
mostly finished diffoscoping, in ossec-server, it's capturing the container hostname so it's not reproducible yet:
```
│ │ ├── ./var/ossec/etc/ossec.conf
│ │ │ @@ -1,13 +1,13 @@
│ │ │  <ossec_config>
│ │ │    <global>
│ │ │      <email_notification>yes</email_notification>
│ │ │      <email_to>root@localhost</email_to>
│ │ │      <smtp_server>127.0.0.1</smtp_server>
│ │ │ -    <email_from>ossecm@be0083033c5f</email_from>
│ │ │ +    <email_from>ossecm@c26ce720f917</email_from>
│ │ │    </global>
```

Also I think I unintentionally renamed the ossec.service unit to ossec-agent.service.

legoktm commented:
The current value is `<email_from>ossecm@focal-sd-generic-ossec-server</email_from>`, I think we can use a fixed hostname like `sd-ossec-builder` (similar to `sd-kernel-builder`).

legoktm commented:
Also I backed out the change to use debhelper's systemd support because it's probably best done after we move to compat 11. So the service units are installed exactly the same as before.

legoktm commented:
Some remaining TODOs (but ready for reviews in absence of that):
* [x] Add CI to verify all debs (except securedrop-app-code for now) are reproducible.
* [x] Document that .buildinfo files need to be published (n.b. this should cover SDW too)

legoktm committed:
Verify all packages except app-code are reproducible in CI

With the builder refactor, we can now run two sets of builds in CI
without it being too slow: at ~14m it finishes before the full app-test
job. And this will get dramatically faster once we have reproducible
wheels.

We run the main SecureDrop and OSSEC package builds twice and then
diffoscope all the debs (except securedrop-app-code). Using reprotest
isn't super useful here since the container would hide many of the
intended changes. But it's a nice TODO for the future. In that sense,
this is mostly proving that in a controlled environment, the builds are
reproducible.

cfm commented on builder/Dockerfile:
```diff
@@ -1,10 +1,9 @@
-# ubuntu:focal-20210921
-FROM ubuntu@sha256:3555f4996aea6be945ae1532fa377c88f4b3b9e6d93531f47af5d78a7d5e3761
-
+FROM ubuntu:focal
```
This is one half of the documentation change @eloquence requested last week: removing <https://developers.securedrop.org/en/latest/dockerbuildmaint.html>.  The other half, per <https://github.com/freedomofpress/securedrop/pull/6754#issuecomment-1444097514>, is documenting the generation of `.buildinfo` files instead.

cfm replied:
This is proposed in freedomofpress/securedrop-dev-docs#40.

cfm reviewed (commented):
I've read through this both commit- and diff-wise and think it looks fabulous, @legoktm.  The precedent and benefit of #6544 is evident!

I'll try to run through the test plan this week, next Monday at the latest.  In the meantime: do you see freedomofpress/securedrop-engineering#20 as a prerequisite for approving (in addition to documenting the goals and principles of) this work?

legoktm commented:
Thanks!

> I'll try to run through the test plan this week, next Monday at the latest. In the meantime: do you see [freedomofpress/securedrop-engineering#20](https://github.com/freedomofpress/securedrop-engineering/pull/20) as a prerequisite for approving (in addition to documenting the goals and principles of) this work?

No. Our current security posture/requirement is that we build debs in a dispVM that builds/executes a container, which doesn't change with this PR (just how the container is set up and what it does), my proposal is to drop the dispVM part, which is what the dep-check stuff is for (since you don't really need to worry about out of date containers in a dispVM), but that should be mostly independent to this.

And I'll work on the documentation stuff in the next few days.

legoktm commented:
Documentation PR: https://github.com/freedomofpress/securedrop-dev-docs/pull/40

cfm reviewed (dismissed):
I've run through a full set of builds according to the test plan, comparing packages built from `0fb3f24^` (as the parent of this branch, to filter out any noise from `develop`) and from this branch.  All look good.  Thanks, @legoktm, for the hints on what level of reproducibility, thus what size and kind of diff, to expect for each packge.

I've received a few OSSEC alerts *about* the new `ossec-server`:

```
Integrity checksum changed for: '/var/ossec/checksdconfig.py'
Permission change: 'rwxr-xr-x' to 'r-xr-x---'
```

```
Integrity checksum changed for: '/var/ossec/etc/ossec.conf'
Permission change: 'rw-r--r--' to 'r--r-----'
```

```
Integrity checksum changed for: '/var/ossec/etc/resolv.conf'
What changed:
1c1
< nameserver 10.139.1.1
---
> -e nameserver 10.139.1.1
```

```
Integrity checksum changed for: '/var/ossec/etc/ossec-init.conf'
What changed:
3c3
< DATE="Thu Mar 11 20:36:21 UTC 2021"
---
> DATE="Mon Apr 13 18:53:21 UTC 2020"
```

I've now reached `Max emails per hour reached.`  Assuming (a) that these diffs are expected, @legoktm, and (b) that alerts come through as usual overnight, I'll merge in the morning.

legoktm commented:
Hmm, the permissions changes are not intentional, let me poke at that tomorrow. (The DATE one is though)

legoktm committed:
Few tweaks to the builder script

* After rebuilding the container re-run the dep-check to reset the
status.
* Create the build/focal directory outside the container.

legoktm committed:
Fix permissions of OSSEC package files to match previous packaging

These aren't necessarily ideal, but reduce the diff as part of this
packaging refactoring.

legoktm committed:
Fix literal "-e" ending up in /var/ossec/etc/resolv.conf

Seems like `make`'s echo is different from bash and doesn't need `-e` to
expand newlines.

legoktm commented:
The easiest way to verify the permission fixes is to use the `debdiff` tool (`devscripts` package in Debian/Fedora). Before today's commits, it looked like https://gist.github.com/legoktm/897ca3a06677385994311a96c8d98aa7 (where original-server is the deb from https://apt.freedom.press/pool/main/o/ossec.net/). It lists files as different if they have different perms/owners/names but ignores content.

After the fixperms fixes, it looks lke:
```
user@dev ~/g/f/s/b/focal> debdiff original-server.deb ossec-server_3.6.0+focal_amd64.deb
[The following lists of changes regard files as different if they have
different names, permissions or owners.]

Files in second .deb but not in first
-------------------------------------
-rw-r--r--  root/root   /usr/share/doc/ossec-server/changelog.gz
-rw-r--r--  root/root   DEBIAN/conffiles
-rw-r--r--  root/root   DEBIAN/md5sums

Files in first .deb but not in second
-------------------------------------
-rw-r--r--  root/root   /usr/share/doc/ossec-server/changelog.Debian
-rw-r--r--  root/root   /usr/share/lintian/overrides/ossec-server
-rw-r--r--  root/root   DEBIAN/templates
-rwxr-xr-x  root/root   DEBIAN/prerm

...
```

All of the remaining changes are expected and desired because they're actually files being added/removed.

cfm reviewed (approved):
Thanks, @legoktm.  I've `debdiff`ed and installed the new `ossec-{agent,server}` packages, and I'll merge tomorrow assuming (a) OSSEC alerts come through as usual overnight and (b) CircleCI comes back from the dead.

cfm reviewed (changes requested):
Oops: can you rebase and resolve the conflicts with `develop` after #6739?

cfm commented:
Rebased from `develop` without changes.

cfm reviewed (approved):
After <https://github.com/freedomofpress/securedrop/pull/6754#pullrequestreview-1340506894>, OSSEC is happy, CI is happy, and the conflicts with `develop` are resolved—so I'm merging.  Thanks for this, @legoktm!
